torch/distributed/fsdp/fully_sharded_data_parallel.py

```diff
@@ -4091,18 +4090,15 @@ def _get_default_comm_hook(self) -> Any:
         Returns a default communication hook based on a sharding strategy.
         """
         if self.sharding_strategy != ShardingStrategy.NO_SHARD:
-            return None
+            return default_hooks.reduce_scatter_hook
```

Comment:
since `reduce_scatter_hook` is now a valid comm hook which can accept `output` to write the result into, we should probably update the signature of hooks accepted by `register_comm_hook` to include this optional function signature and describe how to use it? 

Also, we should clarify the current state of the hook signature we expect:

1. Do we expect different signatures for NO_SHARD vs sharded cases?
2. If we do, is the case where we write into output unsupported for NO_SHARD case, and is the case where we don't write into output and return a `torch.Tensor` unsupported for the sharded case? 

what would it take to achieve unification here?

Reply:
@rohan-varma  let me make sure I understand this:
Regarding question 1: Yes, sharded cases require additional `output` tensor for `reduce_scatter`
question 2: 2.1 we don't return anything, `grad` or `output` are changed in-place. 2.2 It would be feasible to achieve a unification, if `all_reduce` had an `output` argument the way `reduce_scatter` does. Currently, `all_reduce` takes provided `tensor` and changes it in-place. On the other hand, `reduce_scatter` takes `input_tensor` and `output_tensor` and changes `output_tensor` in-place. Since there is difference in logic, `output` is not even available for `NO_SHARD` case, i.e. it won't be initialized and all manipulation with the gradient's tensor are happening in-place. For sharded cases, after `output` gets reduced sum of a gradient's chunk, there is a number of procedures (specific to sharded cases), that involves `output` (e.g. saving it to `param._saved_grad_shard` to support gradient accumulation outside `no_sync()`)